### ISSUE TITLE
Missing key for encryption: %s in PasswordStore

### ISSUE THREAD

**Issue by nyarly** [open-keychain/open-keychain#2588]
The same error as reported in #1781, but removing and re-authorizing the application didn't help.

## Expected Behavior

Should be able to add new keys to PasswordStore

## Current Behavior

PasswordStore reports "OpenKeyChain: missing key for encryption: %s"

## Steps to Reproduce (for bugs)

1. Create a new key in password store
2. Tap "save" icon

## Context

Can't generate new passwords on phone with passwordstore

## Your Environment

* Android Version: 10
* Device Model: Pixel 3
* OpenKeychain Version: 5.5
* From Google Play or F-Droid?: Play

**Comment by hatzel** [comment]
Same here: F-Droid version, Pixel 4A.

**Comment by brogers** [comment]
I am experiencing the same behavior as well.

**Comment by minikN** [comment]
Yes, me too.

**Comment by edvorg** [comment]
Same here

**Comment by scobiehague** [comment]
Same here. F-Droid Galaxy S9 using Yubikey Neo NFC.

**Comment by np2048** [comment]
+1 
 
F-Droid 
OpenKeychain Version: 5.5
Password Store 1.13.2
Android 9 AospExtended-v6.0

**Comment by msfjarvis** [comment]
I have so far never managed to reproduce the problem. The error's thrown [here](https://github.com/open-keychain/open-keychain/blob/258cd4c8367d62515bf71f4a919b3cf67aa04e65/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/PgpSignEncryptOperation.java#L672), which means that the master encryption key could not be found. I don't suspect this is something to be fixed in Password Store since we do not use the account-based API that was seemingly the cause in the other thread.

**Comment by joshua-mcintosh** [comment]
I have had this problem since about when it was first reported, and believe I just figured it out. If you look at the root of the repository, there is a .gpg-id file. The file contained a key ID (eg '01234567FACE'), but not a hexadecimal initializer ('0x'). I removed the file and attempted to save a new key, and it prompted for a recipient and created a new version of the file as well as the test key. So, I may recommend removing the file if you are encountering this to see if it helps.

**Comment by msfjarvis** [comment]
> I have had this problem since about when it was first reported, and believe I just figured it out. If you look at the root of the repository, there is a .gpg-id file. The file contained a key ID (eg '01234567FACE'), but not a hexadecimal initializer ('0x'). I removed the file and attempted to save a new key, and it prompted for a recipient and created a new version of the file as well as the test key. So, I may recommend removing the file if you are encountering this to see if it helps.

Thanks, this was extremely helpful. It seems that in APS we're relatively more tolerant about the leading `0x` than OpenKeychain, so while we let it pass OpenKeychain ends up failing. I'll patch it on our end, but OpenKeychain should still be fixing up that format string to include the invalid key ID.

cc @Valodim

**Comment by nyarly** [comment]
I've just tried something similar - I use password-store on my workstations as well, and have multiple recipients. I added the 0x to my .gpg-ids to see if that would work. There's no drawback, but I still have the %s error. Is it possible that all the recipients are passed as one somehow?

**Comment by s1shed** [comment]
I removed the `~/.password-store/.gpg-id` file from my repo and let Password Store / OpenKeychain create it. 

The key IDs are the same but all in lowercase. Before removing the file the IDs were all in uppercase.

**Comment by nyarly** [comment]
Changing from uppercase to lowercase doesn't correct the issue for me.

Again, I suspect that the multiple IDs cause the problem.

**Comment by s1shed** [comment]
> Changing from uppercase to lowercase doesn't correct the issue for me.

I still had this problem when creating passwords in a subfolder until I removed the .gpg-id from its folder. I currently have a .gpg-id in the root of the repository and now I can use Password Store on multiple devices but the password entries are encrypted with my two keys. It's not optimal but it's good enough for the time being.

**Comment by danjl1100** [comment]
Following the suggestions above (re-authorizing the app, and adding `0x` prefix to `.gpg-id` file) didn't work for me.
I kept getting the error `OpenKeyChain: missing key for encryption: %s`

### Experimenting
I tried creating a fresh repo on the app, and noticed the crucial difference to my existing repo:
* On my Android phone, using Password Store (1.13.5) and OpenKeychain (5.7.5), the created `.gpg-id` references the **Master** key only.
* On my existing repo, I had previously updated `.gpg-id` to reference the **Encryption** subkey.

Changing my existing repo to reference the **Master** key id in `.gpg-id` appears to be working correctly (desktop and Android both using the correct/same key).  I suppose specifying the **Master** key implies that you will use the **Encryption** subkey (?).  Hopefully I don't find out somewhere down the line that my setup is horribly broken  (:

### Verifying
To check, I examined a few existing .gpg files using `cat my-password-file.gpg | gpg --list-packets` and got:
```text
gpg: encrypted with 4096-bit RSA key, ID 0x [... the **Encryption** subkey...]
```
I get the same result on new files (with `.gpg-id` referencing the **Master** key) so it appears everything is working well.


# TL;DR
Fix your existing repo by changing `.gpg-id` to reference the **Master** key (*top yellow bar, stamp picture*, below).

Screenshot of OpenKeychain (5.7.5) app > pick your key > Advanced > Subkeys tab.
![OpenKeychain_master_key_subkey_ids](https://user-images.githubusercontent.com/45136864/162348830-e713bcec-0b29-4b4f-8f39-0ab5a71c2798.jpg)

Using the Encryption sub-key (*blue bar, lock picture*, below) no longer works.  It used to work, but appears this behavior changed at some prior Password Store or OpenKeychain Android app update since ~2019-2021 timeframe.

___

Lastly, I very much appreciate the great work done on OpenKeychain and Password Store Android apps.  I recall setting this up many years ago (~2016?)... and even to this day, it all just works seamlessly!
Thank you :smiley:

**Comment by tennox** [comment]
From reading your finds I had the hunch that the "fix" is actually to have only one entry in `.gpg-id`. Tried it and it fixes the error :+1: 

Only I would like to use a different key on a different machine, and I'm not sure how to get this to work.

I already tried using two keys in different order, suspecting it might just try the first/last one, but to no avail :(

Any ideas?

**Comment by tennox** [comment]
I just figured out the solution for my issue causing `missing key for encryption` (not sure if it's actually the same as yours):

I had to import all public keys from the `.public-keys/` dir (or more specifically, those used in `.gpg-id`) into OpenKeychain app.

So in my case, not really a bug but [PEBKAC](https://en.wiktionary.org/wiki/PEBCAK) :sweat_smile:

**Comment by nyarly** [comment]
@tennox That was exactly the issue I had! With multiple keys, I needed to import them all into OpenKeychain. In retrospect, that's obvious, but it wasn't clear from the error or documentation.

**Comment by theryaz** [comment]
I've been having the exact same issue. Imported all of the public keys into open keychain and it's working!

Output all public keys:

`gpg --output pubkeys.gpg --export`

Copy the file to phone > import keys from file.

**Comment by tennox** [comment]
Related discussion (I need to start searching discussions too, not only issues):
https://github.com/android-password-store/Android-Password-Store/discussions/1304#discussioncomment-342173

**Comment by tennox** [comment]
@nyarly this issue could thous be re-purposed (or rephrased) to add an API to check available keys, as that's why android-password-store can't do anythin about it:
https://github.com/android-password-store/Android-Password-Store/issues/1835#issuecomment-1093263980

**Comment by b-zee** [comment]
I also encountered this issue. I removed the `.gpg-id` file and let the interface select the key. ([From this answer.](https://github.com/open-keychain/open-keychain/issues/2588#issuecomment-778886913))

The new `.gpg-id` that works contains a lowercase hexadecimal with a `0x` prefix. While the old file had uppercase hex without the prefix.

**Comment by akiva** [comment]
I, too, experience this issue. I have an offline master key and I create unique sub-keys for each device I own. While I can decrypt my `pass` database, I cannot encrypt without experiencing this issue due to open-keychain. None of the suggestions above work for me.

**Comment by wolfgangkarall** [comment]
> Changing my existing repo to reference the **Master** key id in `.gpg-id` appears to be working correctly (desktop and Android both using the correct/same key). I suppose specifying the **Master** key implies that you will use the **Encryption** subkey (?). 

I stumbled across this issue when moving one of my kids to a new mobile, and no 0x prefix, nor upper-lowercase helped, but exactly this did: using the master key ID instead of the encryption key ID. Thanks a lot.

**Comment by sgerloff** [comment]
Came across this issue when using Android Password Store together with Open Keychain to sync to a `gopass` repository. The issue was caused by other recipients not being added to the public keyring. The fix included tracking down the repository on the Android device and manually importing the public keys of all recipients to Open Keychain.
